zipline/pipeline/loaders/earnings_estimates.py

```diff
@@ -542,6 +699,332 @@ def get_last_data_per_qtr(self, assets_with_data, columns, dates):
+        if pre_adjustments:
+            adjustment_values, date_indexes = pre_adjustments
+            # We need to undo all adjustments that happen before the
+            # split_asof_date here by reversing the split ratio.
+            adjustments_to_undo = [Float64Multiply(
+                0,
+                split_adjusted_asof_date_idx,
+                sid_idx,
+                sid_idx,
+                1/future_adjustment
+            ) for future_adjustment in adjustment_values]
+            for column_name in requested_split_adjusted_columns:
+                col_to_split_adjustments[column_name] = defaultdict(list)
+                col_to_split_adjustments[column_name][0].extend(
+                    adjustments_to_undo
+                )
+
+                for adjustment, date_index in zip(adjustment_values,
+                                                  date_indexes):
+                    col_to_split_adjustments[
+                        column_name
+                    ][date_index].append(
+                        Float64Multiply(
+                            0,
+                            split_adjusted_asof_date_idx,
+                            sid_idx,
+                            sid_idx,
+                            adjustment
+                        )
+                    )
+        return col_to_split_adjustments
+
+
+    def collect_post_asof_split_adjustments(self,
+                                            post_adjustments,
+                                            requested_qtr_data,
+                                            sid,
+                                            sid_idx,
+                                            sid_estimates,
+                                            requested_split_adjusted_columns):
+        """
+        Collect split adjustments that occur after the
+        split-adjusted-asof-date. Each adjustment needs to be applied to all
+        dates on which knowledge for the requested quarter was older than the
+        date of the adjustment.
+
+        Parameters
+        ----------
+        post_adjustments : tuple(list(float), list(int), pd.DatetimeIndex)
+            The adjustment values, indexes in `dates`, and timestamps for
+            adjustments that happened after the split-asof-date.
+        requested_qtr_data : pd.DataFrame
+            The requested quarter data for each calendar date per sid.
+        sid : int
+            The sid for which adjustments need to be collected.
+        sid_idx : int
+            The index of `sid` in the adjusted array.
+        requested_split_adjusted_columns : list of str
+            The requested split adjusted columns.
+        Returns
+        -------
+        col_to_split_adjustments : dict[str -> dict[int -> list of Adjustment]]
+            The adjustments for this sid that occurred after the
+            split-asof-date.
+        """
+        col_to_split_adjustments = defaultdict(dict)
```

Comment:
My recommendation would be to not use `defaultdict`.
It is a very powerful too, but has historically been the root cause for many bugs and/or memory leaks.

Instead of:
```
for column_name in requested_split_adjusted_columns:
                            col_to_split_adjustments[
                                column_name
                            ][start_idx].append(
                                Float64Multiply(
                                    # Always apply from first day of qtr
                                    qtr_range[0],
                                    end_idx,
                                    sid_idx,
                                    sid_idx,
                                    adjustment
                                )
                            )
```
The append could be changed to the following:
```
for column_name in requested_split_adjusted_columns:
    adj = Float64Multiply(
      # Always apply from first day of qtr
      qtr_range[0],
      end_idx,
      sid_idx,
      sid_idx,
      adjustment)
   split_adjustments = col_to_split_adjustments[column_name]
   try:
        adjs_at_index[start_index].append(adj)
   except KeyError:
        adjs_at_index[start_index] = [adj]
```

By avoiding `defaultdict` a scenario is avoided where a consumer is iterating over candidate `start_idx` values and quietly creating an empty list for each value, when a KeyError should be raised.